python/lsst/daf/butler/core/_butlerUri.py

```diff
@@ -0,0 +1,1332 @@
+    paths = list(posixPath.parts)
+
+    # Have to convert the root directory after splitting
+    if paths[0] == posixPath.root:
+        paths[0] = OS_ROOT_PATH
+
+    # Trailing "/" is stripped so we need to add back an empty path
+    # for consistency
+    if str(posix).endswith(posixpath.sep):
+        paths.append("")
+
+    return os.path.join(*paths)
+
+
+class NoTransaction:
+    """A simple emulation of the `DatastoreTransaction` class.
+
+    Does nothing.
+    """
+
+    def __init__(self) -> None:
+        return
+
+    @contextlib.contextmanager
+    def undoWith(self, name: str, undoFunc: Callable, *args: Any, **kwargs: Any) -> Iterator[None]:
+        """No-op context manager to replace `DatastoreTransaction`
+        """
+        yield None
+
+
+class ButlerURI:
+    """Convenience wrapper around URI parsers.
+
+    Provides access to URI components and can convert file
+    paths into absolute path URIs. Scheme-less URIs are treated as if
+    they are local file system paths and are converted to absolute URIs.
+
+    A specialist subclass is created for each supported URI scheme.
+
+    Parameters
+    ----------
+    uri : `str` or `urllib.parse.ParseResult`
+        URI in string form.  Can be scheme-less if referring to a local
+        filesystem path.
+    root : `str`, optional
+        When fixing up a relative path in a ``file`` scheme or if scheme-less,
+        use this as the root. Must be absolute.  If `None` the current
+        working directory will be used.
+    forceAbsolute : `bool`, optional
+        If `True`, scheme-less relative URI will be converted to an absolute
+        path using a ``file`` scheme. If `False` scheme-less URI will remain
+        scheme-less and will not be updated to ``file`` or absolute path.
+    forceDirectory: `bool`, optional
+        If `True` forces the URI to end with a separator, otherwise given URI
+        is interpreted as is.
+    """
+
+    _pathLib: Type[PurePath] = PurePosixPath
+    """Path library to use for this scheme."""
+
+    _pathModule = posixpath
+    """Path module to use for this scheme."""
+
+    transferModes: Tuple[str, ...] = ("copy", "auto", "move")
+    """Transfer modes supported by this implementation.
+
+    Move is special in that it is generally a copy followed by an unlink.
+    Whether that unlink works depends critically on whether the source URI
+    implements unlink. If it does not the move will be reported as a failure.
+    """
+
+    transferDefault: str = "copy"
+    """Default mode to use for transferring if ``auto`` is specified."""
+
+    quotePaths = True
+    """True if path-like elements modifying a URI should be quoted.
+
+    All non-schemeless URIs have to internally use quoted paths. Therefore
+    if a new file name is given (e.g. to updateFile or join) a decision must
+    be made whether to quote it to be consistent.
+    """
+
+    # mypy is confused without this
+    _uri: urllib.parse.ParseResult
```

Comment:
You would think that it would be enough to declare what type `parsed` was but if I leave this line out it really doesn't like it. It must be the use of `__new__`.